### PR TITLE
liballoc: VecDeque: Add binary search functions

### PR DESCRIPTION
**Pull request by vojtechkral**
I am submitting rust-lang/rfcs#2997 as a PR as suggested by @scottmcm 

~~I haven't yet created a tracking issue - if there's a favorable feedback I'll create one and update the issue links in the unstable attribs.~~

Tracking issue: #78021

### CODE REVIEW

**Review by dtolnay — APPROVED** [review]
Thanks, this is great.

Please make a tracking issue and update the stability attributes in the PR.

File: library/alloc/src/collections/vec_deque.rs [hunk 1]
```diff
@@ -2430,6 +2430,154 @@ impl<T> VecDeque<T> {
...
+    /// assert_eq!(deque.binary_search_by(|x| x.cmp(&100)), Err(13));
+    /// let r = deque.binary_search_by(|x| x.cmp(&1));
+    /// assert!(matches!(r, Ok(1..=4)));
+    /// ```
+    #[unstable(feature = "vecdeque_binary_search", issue = "1")]
+    pub fn binary_search_by<'a, F>(&'a self, mut f: F) -> Result<usize, usize>
+    where
+        F: FnMut(&'a T) -> Ordering,
+    {
+        if self.is_empty() {
```
[the remark this answers is not in the document]
Comment (vojtechkral) [reply library/alloc/src/collections/vec_deque.rs:2519]: Yeah, probably not...

**Comment by vojtechkral** [comment]
Added a tracking issue, simplified implementation as suggested by @scottmcm ...

**Comment by scottmcm** [comment]
Stability attributes and code updates look good to me; thanks!

@bors r=scottmcm,dtolnay

(Feel free to `r-` again if needed, but based on your review being a ✔ I assume you're ok with this.)

**Pull request merged**